### PR TITLE
feat: bitand, bitor, shl, shr, arithmetic_shr

### PR DESCRIPTION
**Pull request by freesig**
Adds some basic bitwise and shifting operations.

### CODE REVIEW

File: crates/constraint-vm/src/alu.rs [hunk 1]
```diff
@@ -22,6 +25,33 @@ pub(crate) fn mod_(a: Word, b: Word) -> OpResult<Word> {
...
+}
+
+pub(crate) fn arithmetic_shr(a: Word, b: Word) -> OpResult<Word> {
+    check_shift_bounds(b)?;
+    Ok(a >> b)
+}
+
+const BITS_IN_WORD: Word = core::mem::size_of::<Word>() as Word * 8;
+
+#[inline]
```
Comment (mitchmindtree) [inline crates/constraint-vm/src/alu.rs:46]: rustc will probs do this anyway, but probs won't hurt either

File: crates/constraint-vm/src/alu.rs [hunk 2]
```diff
@@ -22,6 +25,33 @@ pub(crate) fn mod_(a: Word, b: Word) -> OpResult<Word> {
...
+fn check_shift_bounds(b: Word) -> OpResult<()> {
+    let bounds = 0..BITS_IN_WORD;
+    if !bounds.contains(&b) {
+        return Err(AluError::Overflow.into());
+    }
+    Ok(())
+}
```
Comment (mitchmindtree) [inline crates/constraint-vm/src/alu.rs:53]: Another option might be to use `.checked_shl/shr` which I think does this for us

File: crates/types/src/lib.rs [hunk 3]
```diff
@@ -69,7 +69,7 @@ pub struct PredicateAddress {
 /// A protocol block.
 pub struct Block {
     /// The block number.
-    pub number: u64,
+    pub number: Word,
```
Comment (mitchmindtree) [inline crates/types/src/lib.rs:72]: lol nice

File: crates/constraint-vm/src/alu.rs [hunk 4]
```diff
@@ -22,6 +25,33 @@ pub(crate) fn mod_(a: Word, b: Word) -> OpResult<Word> {
...
 
+pub(crate) fn shl(a: Word, b: Word) -> OpResult<Word> {
+    check_shift_bounds(b)?;
+    Ok(a << b)
+}
+
+pub(crate) fn shr(a: Word, b: Word) -> OpResult<Word> {
+    check_shift_bounds(b)?;
+    // casts are safe and turn this into a logical shift
+    Ok(((a as u64) >> b) as i64)
```
Comment (mitchmindtree) [inline crates/constraint-vm/src/alu.rs:36]:
```suggestion
    Ok(((a as u64) >> b) as Word)
```

**Review by mitchmindtree — APPROVED** [review]
Nice

File: crates/constraint-vm/src/alu.rs (hunk 2 above)
Comment (freesig) [reply crates/constraint-vm/src/alu.rs:53]: Yeh I tried that but they all take other datatypes on the rhs like u32. This forces you to do a try into then also call checked. Where as this way there's just a single check

**Pull request merged**